### PR TITLE
[WebAssembly] support getVT from `externref` and `funcref`

### LINKED ISSUE #69894: Compilation crash targetting wasm with -O2 and reference types
Tested against:

Ubuntu clang version 17.0.3 (++20231017073226+888437e1b600-1~exp1~20231017073338.54)
clang version 18.0.0 (https://github.com/llvm/llvm-project d1e685df45dc5944b43d2547d0138cd4a3ee4efe)

cc @sbc100 @pmatos

### testcase.c

```c
__externref_t foo(void);
void bar(__externref_t);

void
test(int flag, __externref_t ref1, __externref_t ref2)
{
  if (flag) {
    ref1 = foo();
    ref2 = foo();
  }
  bar(ref1);
  bar(ref2);
}
```
Compilation command:
```sh
clang -c testcase.c -O2   -mreference-types -target wasm32-unknown-unknown
```

The files it asked me to upload:
### testcase-ccc7e1.c

```
# 1 "<built-in>"
# 1 "testcase.c"
__externref_t foo(void);
void bar(__externref_t);

void
test(int flag, __externref_t ref1, __externref_t ref2)
{
  if (flag) {
    ref1 = foo();
    ref2 = foo();
  }
  bar(ref1);
  bar(ref2);
}
```

### testcase-ccc7e1.sh

```
# Crash reproducer for Ubuntu clang version 17.0.3 (++20231017073226+888437e1b600-1~exp1~20231017073338.54)
# Driver args: "-c" "testcase.c" "-O2" "-mreference-types" "-target" "wasm32-unknown-unknown"
# Original command:  "/usr/lib/llvm-17/bin/clang" "-cc1" "-triple" "wasm32-unknown-unknown" "-emit-obj" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "testcase.c" "-mrelocation-model" "static" "-mframe-pointer=none" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-target-cpu" "generic" "-target-feature" "+reference-types" "-fvisibility=hidden" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/home/hood/Documents/programming/pyodide" "-resource-dir" "/usr/lib/llvm-17/lib/clang/17" "-internal-isystem" "/usr/lib/llvm-17/lib/clang/17/include" "-O2" "-fdebug-compilation-dir=/home/hood/Documents/programming/pyodide" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fcolor-diagnostics" "-vectorize-loops" "-vectorize-slp" "-o" "testcase.o" "-x" "c" "testcase.c"
 "/usr/lib/llvm-17/bin/clang" "-cc1" "-triple" "wasm32-unknown-unknown" "-emit-obj" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "testcase.c" "-mrelocation-model" "static" "-mframe-pointer=none" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-target-cpu" "generic" "-target-feature" "+reference-types" "-fvisibility=hidden" "-debugger-tuning=gdb" "-fcoverage-compilation-dir=/home/hood/Documents/programming/pyodide" "-O2" "-fdebug-compilation-dir=/home/hood/Documents/programming/pyodide" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fcolor-diagnostics" "-vectorize-loops" "-vectorize-slp" "-x" "c" "testcase-ccc7e1.c"
```

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixes: #69894 
#71069 try to treat `externref` as invalid element type but It should be a normal opaque pointer containing some unknown data.
`getVT` should be inverse operation of `getTypeForEVT` which convert `MVT::externref` to `externref`.

### CODE REVIEW

File: clang/test/CodeGen/WebAssembly/wasm-externref-novec.c [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -O2 -triple wasm32-unknown-unknown-wasm -target-feature +reference-types -emit-llvm -o - %s | FileCheck %s
```
Comment (aheejin) [inline clang/test/CodeGen/WebAssembly/wasm-externref-novec.c:1]:
```suggestion
// RUN: %clang_cc1 -O2 -triple wasm32-unknown-unknown -target-feature +reference-types -emit-llvm -o - %s | FileCheck %s
```

Comment (aheejin) [reply clang/test/CodeGen/WebAssembly/wasm-externref-novec.c:1]: By the way, this command succeeds even without this patch. The test should be something that fails now and succeeds with the fix.

File: clang/test/CodeGen/WebAssembly/wasm-externref-novec.c [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+// From issue 69894. Reftypes need to be marked as not valid as vector elements.
```
Comment (aheejin) [inline clang/test/CodeGen/WebAssembly/wasm-externref-novec.c:3]: If you would like to refer to an issue, I think including the URL is more convenient: `https://github.com/llvm/llvm-project/issues/69894`

**Review by aheejin — COMMENTED** [review]
What happens if you compile a file that really vectorizes reference types? The attached test doesn't seem to actually end up getting vectorized after all. Can you add a test that get vectorized?

**Comment by HerrCai0907** [comment]
> What happens if you compile a file that really vectorizes reference types? The attached test doesn't seem to actually end up getting vectorized after all. Can you add a test that get vectorized?

Is it possible to vectorize external reference types? I don't think there are valid test case for it.

**Pull request closed without merging**